### PR TITLE
Fix backspace always unfolding previous line

### LINKED ISSUE #89522: Using backspace in script editor unfolds folded region on previous line
### Tested versions

Version: v4.3.dev4.official [df78c0636]

### System information

Windows 10, Geforce 960, Vulkan

### Issue description

In the script editor, pressing backspace to remove a character on the line after a block of folded code unfolds the code. This does not occur with other text-deletion methods that I tried, including delete, ctrl+backspace, or using backspace on some highlighted text. 

### Steps to reproduce

First:
image

Then after hitting backspace:
image

### Minimal reproduction project (MRP)

(hopefully unnecessary)

### PR DESCRIPTION
**Pull request by smntic**
Add a condition to `CodeEdit::_backspace_internal` so the previous line is only unfolded when the caret is at the start of the current line.

Fixes #89522

### CODE REVIEW

**Review by Paulb23 — COMMENTED** [review]
Changes looks good, would be ideal to add a unit test to capture this behaviour.

File: tests/scene/test_code_edit.h [hunk 1]
```diff
@@ -3972,6 +3972,24 @@ TEST_CASE("[SceneTree][CodeEdit] Backspace delete") {
 	code_edit->backspace();
 	CHECK(code_edit->get_text() == "line 1\nline 2\nline 3");
 
+	SUBCASE("[CodeEdit] folding") {
+		code_edit->set_line_folding_enabled(true);
+
+		/* Unfold previous folded line on backspace if caret is at the first column. */
+		code_edit->set_text("");
+		code_edit->insert_text_at_caret("line 1\n\tline 2\n");
```
Comment (kitbdev) [inline tests/scene/test_code_edit.h:3980]:
```suggestion
		code_edit->set_text("line 1\n\tline 2\nline 3");
		code_edit->set_caret_line(2);
		code_edit->set_caret_column(0);
```

I think something like this is more clear.

File: tests/scene/test_code_edit.h [hunk 2]
```diff
@@ -3972,6 +3972,24 @@ TEST_CASE("[SceneTree][CodeEdit] Backspace delete") {
...
+		code_edit->fold_line(0);
+		code_edit->backspace();
+		CHECK_FALSE(code_edit->is_line_folded(0));
+
+		/* Do not unfold previous line on backspace if the caret is not at the first column. */
+		code_edit->set_text("");
+		code_edit->insert_text_at_caret("line 1\n\tline 2\nline 3");
```
Comment (kitbdev) [inline tests/scene/test_code_edit.h:3987]:
```suggestion	
		code_edit->set_text("line 1\n\tline 2\nline 3");
		code_edit->set_caret_line(2);
		code_edit->set_caret_column(4);
```
Same as above.

File: tests/scene/test_code_edit.h [hunk 3]
```diff
@@ -3972,6 +3972,24 @@ TEST_CASE("[SceneTree][CodeEdit] Backspace delete") {
 	code_edit->backspace();
 	CHECK(code_edit->get_text() == "line 1\nline 2\nline 3");
 
+	SUBCASE("[CodeEdit] folding") {
+		code_edit->set_line_folding_enabled(true);
+
+		/* Unfold previous folded line on backspace if caret is at the first column. */
```
Comment (kitbdev) [inline tests/scene/test_code_edit.h:3978]:
```suggestion
		// Unfold previous folded line on backspace if caret is at the first column.
```
Single line comments are preferred, even though block comments are used elsewhere in this file.
Same with below.

File: scene/gui/code_edit.cpp [hunk 4]
```diff
@@ -718,7 +718,7 @@ void CodeEdit::_backspace_internal(int p_caret) {
 			continue;
 		}
 
-		if (cl > 0 && _is_line_hidden(cl - 1)) {
+		if (cl > 0 && cc == 0 && _is_line_hidden(cl - 1)) {
```
Comment (kitbdev) [inline scene/gui/code_edit.cpp:721]: Needs rebase after #86978,
`cc` was renamed to `to_column`.

**Review by Paulb23 — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 4.3.1.

**Pull request merged**